src/main/java/build/buildfarm/worker/OutputDirectory.java

```diff
@@ -0,0 +1,89 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.worker;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+// would be great to be able to inherit from ImmutableMap...
+public class OutputDirectory extends HashMap<String, OutputDirectory> {
+  private OutputDirectory() {
+  }
+
+  public static OutputDirectory parse(Iterable<String> outputFiles, Iterable<String> outputDirs) {
+    return parseDirectories(Iterables.mergeSorted(
```

Comment:
Wait, now you no longer need to sort them here, right?
In fact, I think if you add a:
Iterables.transform(outputDirs, (d) -> d.charAt(d.length() - 1) == '/') ? d : d + "/")
here and leave the mergeSorted, then you don't have to do anything extra in the next function, and the code will be more obviously correct (everything has a / or is an empty string, no assumptions on input).

Reply:
I will still need to sort the transformed Iterable - `mergeSorted` is predicated on non-descending order for the set of Iterables

Reply:
Yeah, the `mergeSorted` is no longer applicable with your correct observation that the initial sorting of output files/dirs is not sufficient for sorting 'parent directory comes first'. This needs to become a `concat`, and everything has a / is my invariant (no empty strings)